# ###############################################################################
# This file is part of sORTholog.                                               #
#                                                                               #
# Authors: Geoffrey Hutinet and Remi Denise                                     #
# Copyright Â© 2021 University of Forida (Gainesville, Florida).                 #
# See the COPYRIGHT file for details.                                           #
#                                                                               #
# sORTholog is a Snakemake workflow that produce a table visualization          #
# of the presence and absence of encoded proteins in a given set of genomes.    #
# The primary output will be the table of sORTholog protein data filled using   #
# the NCBI Protein Database, as well as PDF and PNG files of the table          #
# visualization.                                                                #
#                                                                               #
# sORTholog is free software: you can redistribute it and/or modify             #
# it under the terms of the Affero GNU General Public License as published by   #
# the Free Software Foundation, either version 3 of the License,                #
# or (at your option) any later version.                                        #
#                                                                               #
# sORTholog is distributed in the hope that it will be useful, but              #
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY    #
# or FITNESSFOR A PARTICULAR PURPOSE. See the Affero GNU General Public License #
# for more details.                                                             #
#                                                                               #
# You should have received a copy of the Affero GNU General Public License      #
# along with sORTholog (COPYING file).                                          #
# If not, see <https://www.gnu.org/licenses/>.                                  #
# ###############################################################################

##########################################################################
##########################################################################
##
##                            Singularity
##
##########################################################################
##########################################################################
# this container defines the underlying OS for each job when using the workflow
# with --use-conda --use-singularity


container: "docker://continuumio/miniconda3"


from snakemake.utils import min_version

# set minimum snakemake version
min_version("6.14.0")

##########################################################################
##########################################################################
##
##                                Rules
##
##########################################################################
##########################################################################


# load config
configfile: "config/config.yaml"


# Path to report rst file
report: "report/workflow.rst"


# Set local rule that could be launch without submiting to cluster because ultra quick
localrules:
    all,
    clean,
    merge_fasta,


##########################################################################
# Load rules
##########################################################################


include: "rules/common.smk"
include: "rules/fetch_prot.smk"
include: "rules/format_tables.smk"
include: "rules/ncbi_blast.smk"
include: "rules/plots.smk"
include: "rules/silix.smk"
include: "rules/make_fasta.smk"
include: "rules/analysis_threshold.smk"


##########################################################################
##########################################################################


rule all:
    input:
        plots=get_final_output(),
    log:
        os.path.join(OUTPUT_FOLDER, "logs", "all.log"),


##########################################################################
##########################################################################


rule clean:
    input:
        os.path.join(OUTPUT_FOLDER, "database", "all_taxid"),
        os.path.join(OUTPUT_FOLDER, "database", "merge_fasta"),
        os.path.join(OUTPUT_FOLDER, "database", "reduce_taxid"),
        os.path.join(OUTPUT_FOLDER, "processing_files"),
        os.path.join(OUTPUT_FOLDER, "logs"),
    log:
        os.path.join(OUTPUT_FOLDER, "logs", "clean.log"),
    shell:
        """rm -rf {input}"""


##########################################################################
##########################################################################
